File: prettier-plugin-csharpier/src/Printer/Types/FixedStatement.ts
import { PrintMethod } from "../PrintMethod";
import { SyntaxToken, printPathSyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";

export interface FixedStatementNode extends SyntaxTreeNode<"FixedStatement"> {
    fixedKeyword: SyntaxToken;
    declaration: SyntaxTreeNode;
    statement: SyntaxTreeNode;
}

export const printFixedStatement: PrintMethod<FixedStatementNode> = (path, options, print) => {
    return concat([
        printPathSyntaxToken(path, "fixedKeyword"),
        " ",
        "(",
        path.call(print, "declaration"),
        ")",
        path.call(print, "statement")
    ]);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/FixedStatement.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class FixedStatement
{
    public static Doc Print(FixedStatementSyntax node, FormattingContext context)
    {
        return Doc.Concat(
            ExtraNewLines.Print(node),
            Doc.Group(
                Token.Print(node.FixedKeyword, context),
                " ",
                Token.Print(node.OpenParenToken, context),
                Doc.Group(
                    Doc.Indent(Doc.SoftLine, Node.Print(node.Declaration, context)),
                    Doc.SoftLine
                ),
                Token.Print(node.CloseParenToken, context),
                Doc.IfBreak(Doc.Null, Doc.SoftLine)
            ),
            node.Statement is BlockSyntax blockSyntax
              ? Block.Print(blockSyntax, context)
              : Doc.IndentIf(
                    node.Statement is not FixedStatementSyntax,
                    Doc.Concat(Doc.HardLine, Node.Print(node.Statement, context))
                )
        );
    }
}
